### PR TITLE
CLICKHOUSE-3832: Add HTTP Basic authentification in replication protocol

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/programs/server/Server.cpp [hunk 1]
```diff
@@ -230,6 +230,17 @@ int Server::main(const std::vector<std::string> & /*args*/)
         global_context->setInterserverIOAddress(this_host, port);
     }
 
+    if (config().has("interserver_http_credentials.user"))
+    {
+        String user = config().getString("interserver_http_credentials.user", "");
+        String password = config().getString("interserver_http_credentials.password", "");
+
+        if (user.empty())
+            throw Exception("Empty interserver_http_credentials user can't be empty");
```
Comment (alexey-milovidov) [inline dbms/programs/server/Server.cpp]: Error code

File: dbms/programs/server/Server.cpp [hunk 2]
```diff
@@ -230,6 +230,17 @@ int Server::main(const std::vector<std::string> & /*args*/)
         global_context->setInterserverIOAddress(this_host, port);
     }
 
+    if (config().has("interserver_http_credentials.user"))
```
Comment (alexey-milovidov) [inline dbms/programs/server/Server.cpp]: Maybe `config().has("interserver_http_credentials")` is better.

File: dbms/programs/server/InterserverIOHTTPHandler.cpp [hunk 3]
```diff
@@ -65,8 +92,18 @@ void InterserverIOHTTPHandler::handleRequest(Poco::Net::HTTPServerRequest & requ
 
     try
     {
-        processQuery(request, response);
-        LOG_INFO(log, "Done processing query");
+        if (auto [msg, success] = checkAuthentication(request); success)
```
Comment (alexey-milovidov) [inline dbms/programs/server/InterserverIOHTTPHandler.cpp]: msg is unused.

Comment (alesapin) [reply dbms/programs/server/InterserverIOHTTPHandler.cpp]: it's visible inside else branch.

**Pull request merged**